File: packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/InterceptingWebViewClient.java

package io.flutter.plugins.webviewflutter;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

import android.annotation.TargetApi;
import android.os.Build;
import android.graphics.Bitmap;
import android.view.KeyEvent;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebResourceError;

import okhttp3.CookieJar;
import okhttp3.Cookie;
import okhttp3.OkHttpClient;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.HttpUrl;
import okhttp3.Call;


/** WebViewClient that intercepts requests and injects javascript to all frames **/
public class InterceptingWebViewClient extends WebViewClient {

    private FlutterWebViewClient delegate;
    static MediaType jsonContentType = MediaType.parse("application/json; charset=UTF-8");
    private List<UserScript> scripts;
    private OkHttpClient okHttpClient;
    private CookieJar cookies;

    public InterceptingWebViewClient(FlutterWebViewClient delegate, final List<UserScript> scripts) {
        this.delegate = delegate;
        this.scripts = scripts;
        this.cookies = new CookieJar() {
            Map<String, Map<String, Cookie>> map = new HashMap<>();

            @Override
            public void saveFromResponse(HttpUrl url, List<Cookie> cookies) {
                for (Cookie c : cookies) {
                    Map<String, Cookie> cc = map.get(c.domain());
                    if (cc == null) {
                        cc = new HashMap<>();
                        map.put(c.domain(), cc);
                    }
                    cc.put(c.name(), c);
                }
            }

            @Override
            public List<Cookie> loadForRequest(HttpUrl url) {
                Map<String, Cookie> cc = map.get(url.topPrivateDomain());
                List<Cookie> cookies = new ArrayList<>();
                if (cc == null)
                    return cookies;

                for (Cookie cookie : cc.values()) {
                    if (cookie.hostOnly() && !url.host().equals(cookie.domain()))
                        continue;
                    else if (cookie.httpOnly() && !url.scheme().startsWith("http"))
                        continue;
                    else if (cookie.secure() && !url.isHttps())
                        continue;
                    else if (cookie.expiresAt() < System.currentTimeMillis())
                        continue;
                    else if (!cookie.path().equals("/") && !cookie.path().equals(url.encodedPath()))
                        continue;
                    else
                        cookies.add(cookie);
                }

                return cookies;
            }
        };

        this.okHttpClient = new OkHttpClient.Builder().followRedirects(true).followSslRedirects(true)
                .cookieJar(this.cookies).build();
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    @Override
    public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
        return okhttpFetch(request);
    }

    @TargetApi(Build.VERSION_CODES.N)
    @Override
    public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
        return delegate.shouldOverrideUrlLoading(view, request);
    }

    @Override
    public void onPageStarted(WebView view, String url, Bitmap favicon) {
        delegate.onPageStarted(view, url);
    }

    @Override
    public void onPageFinished(WebView view, String url) {
        delegate.onPageFinished(view, url);
    }

    @TargetApi(Build.VERSION_CODES.M)
    @Override
    public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
        delegate.onWebResourceError(error.getErrorCode(), error.getDescription().toString(),
                request.getUrl().toString());
    }

    @Override
    public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
        delegate.onWebResourceError(errorCode, description, failingUrl);
    }

    @Override
    public void onUnhandledKeyEvent(WebView view, KeyEvent event) {
        // Deliberately empty. Occasionally the webview will mark events as having
        // failed to be
        // handled even though they were handled. We don't want to propagate those as
        // they're not
        // truly lost.
    }

    private WebResourceResponse okhttpFetch(WebResourceRequest request) {
        MediaType contentType = jsonContentType;

        if (request == null || request.getUrl() == null || request.hasGesture()) {
            return null;
        }

        String url = request.getUrl().toString();
        String scheme = request.getUrl().getScheme().toLowerCase().trim();
        String method = request.getMethod() == null ? "GET" : request.getMethod();

        if (!scheme.equals("http") && !scheme.equals("https")) {
            return null;
        }        

        Request.Builder b = new Request.Builder().url(url);

        // Iterate headers. Extract request body if it exists

        for (Map.Entry<String, String> entry : request.getRequestHeaders().entrySet()) {

            String key = entry.getKey().trim().toLowerCase();
            String value = entry.getValue().trim().toLowerCase();

            if (key.equals("accept") && !value.contains("html")) {
                return null;
            } else {
                b = b.addHeader(entry.getKey(), entry.getValue());
            }
        }

        b.method(method, null);

        Request req = b.build();
        
        Call call = okHttpClient.newCall(req);
        return new NonBlockingWebResourceResponse(call, scripts);
    }
}

File: packages/webview_flutter/android/src/main/kotlin/io/flutter/plugins/webviewflutter/InterceptingWebViewClient.kt
package io.flutter.plugins.webviewflutter

import android.annotation.TargetApi
import android.graphics.Bitmap
import android.os.Build
import android.view.KeyEvent
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import android.webkit.WebView
import android.webkit.WebViewClient
import okhttp3.MediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody

/** WebViewClient that intercepts requests and injects javascript to all frames  */
internal class InterceptingWebViewClient(val delegate: FlutterWebViewClient, val scripts: List<UserScript>, val postData: MutableMap<String, String>) : WebViewClient() {

    private val okHttpClient: OkHttpClient
    private val cookies = FlutterCookieManager

    companion object {
        var jsonContentType = MediaType.parse("application/json; charset=UTF-8")!!
    }

    init {
        okHttpClient = OkHttpClient.Builder().followRedirects(true).followSslRedirects(true)
            .cookieJar(cookies).build()

        okHttpClient.dispatcher().setMaxRequestsPerHost(20)
    }

    override fun shouldInterceptRequest(view: WebView?, request: WebResourceRequest): WebResourceResponse? {
        return okhttpFetch(request)
    }

    override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
        delegate.onPageStarted(view, url)
    }

    override fun onPageFinished(view: WebView, url: String) {
        delegate.onPageFinished(view, url)
    }

    @TargetApi(Build.VERSION_CODES.M)
    override fun onReceivedError(view: WebView, request: WebResourceRequest, error: WebResourceError) {
        delegate.onWebResourceError(
            error.getErrorCode(), error.getDescription().toString(),
            request.getUrl().toString()
        )
    }

    override fun onReceivedError(view: WebView, errorCode: Int, description: String?, failingUrl: String?) {
        delegate.onWebResourceError(errorCode, description, failingUrl)
    }

    override fun onUnhandledKeyEvent(view: WebView, event: KeyEvent?) {
        // Deliberately empty. Occasionally the webview will mark events as having
        // failed to be
        // handled even though they were handled. We don't want to propagate those as
        // they're not
        // truly lost.
    }

    private fun okhttpFetch(request: WebResourceRequest?): WebResourceResponse? {

        // System.out.println("OkHTTP intercepted: " + request?.getUrl() ?: "")
        // System.out.println("OkHTTP intercepted method: " + request?.getMethod() ?: "")

        if (request == null || request.getUrl() == null || request.hasGesture()) {
            return null
        }

        // System.out.println("OkHTTP: " + request.getUrl().toString())
        // System.out.println("OkHTTP method: " + request.getMethod())
        val url: String = request.getUrl().toString()
        val scheme: String = request.getUrl().getScheme().toLowerCase().trim()
        val method = (request.getMethod()?.toUpperCase() ?: "GET").trim()

        if (!scheme.equals("http") && !scheme.equals("https")) {
            return null
        }
        var builder = Request.Builder().url(url)

        var contentType = jsonContentType
        var content: String? = null
        var request_id: String? = null

        // Iterate headers. Extract request body if it exists
        for (entry in request.getRequestHeaders().entries) {
            val key = entry.key.trim().toLowerCase()
            val value = entry.value.trim().toLowerCase()

            when (key) {
                "content-type" -> {
                    contentType = MediaType.parse(value) ?: contentType
                    builder.addHeader(entry.key, entry.value)
                }
                "x-cense-request-id" -> {

                    content = postData.get(value)
                    postData.remove(key)
                }
                else -> builder.addHeader(entry.key, entry.value)
            }
        }

        when {
            method == "POST" -> builder.method(method, RequestBody.create(contentType, content ?: "{}"))
            else -> when (content) {
                null -> builder.method(method, null)
                else -> builder.method(method, RequestBody.create(contentType, content))
            }
        }

        val req = builder.build()
        val call = okHttpClient.newCall(req)

        // System.out.println("Intercepting: " + request.getUrl().toString() + " post data: " + content + ". method: " + method)
        return NonBlockingWebResourceResponse(request, call, scripts, postData)
    }
}
